http-server/src/main/java/io/airlift/http/server/HttpServer.java

```diff
@@ -575,29 +508,24 @@ void join()
         server.join();
     }
 
-    private static Set<X509Certificate> loadAllX509Certificates(HttpServerConfig config)
+    /**
+     * Gets the sslContextFactory from the optional if one exists, otherwise builds one using the HttpServerConfig.
+     *
+     * @param maybeSslContextFactory
+     * @param config
+     * @return
+     */
+    private SslContextFactory.Server getSslContextFactory(Optional<SslContextFactory.Server> maybeSslContextFactory, HttpServerConfig config)
     {
-        ImmutableSet.Builder<X509Certificate> certificates = ImmutableSet.builder();
-        if (config.isHttpsEnabled()) {
-            try (InputStream keystoreInputStream = new FileInputStream(config.getKeystorePath())) {
-                KeyStore keystore = KeyStore.getInstance(KeyStore.getDefaultType());
-                keystore.load(keystoreInputStream, config.getKeystorePassword().toCharArray());
-
-                for (String alias : list(keystore.aliases())) {
-                    try {
-                        Certificate certificate = keystore.getCertificate(alias);
-                        if (certificate instanceof X509Certificate) {
-                            certificates.add((X509Certificate) certificate);
-                        }
-                    }
-                    catch (KeyStoreException ignored) {
-                    }
-                }
-            }
-            catch (Exception ignored) {
-            }
+        Supplier<SslContextFactory.Server> supplier;
+        if (maybeSslContextFactory.isPresent()) {
+            supplier = new SingletonSslContextFactorySupplier(maybeSslContextFactory.get());
         }
-        return certificates.build();
+        else {
+            supplier = new ReloadableSslContextFactorySupplier(config, scheduledExecutorService);
+        }
+
+        return supplier.get();
```

Comment:
The purpose of `ReloadableSslContextFactorySupplier` is that it returns a supplier where the `get()` method can return a different factory whenever the keys change. Since we only call the `get()` method once, it seems like this won't work. I think we need to tell Jetty to reload the factory.

Reply:
The SslContextFactory is responsible for always returning an SslContext that has been built using the latest keys. The Jetty APIs that depend on SslContextFactory will only take a single instance, they won't invoke this supplier again to get new instances. Instead they call the factory to get what is assumed to be the correct SslContext when one is needed. 

As an example: The SslConnectionFactory calls [SslContextFactory.newSSLEngine](https://github.com/eclipse/jetty.project/blob/81c46638cd63e34e06f1c99b514d9d15713a8e7b/jetty-server/src/main/java/org/eclipse/jetty/server/SslConnectionFactory.java#L150) every time a connection is established. This captures uses whatever the latest provided [SslContext](https://github.com/eclipse/jetty.project/blob/jetty-10.0.x/jetty-util/src/main/java/org/eclipse/jetty/util/ssl/SslContextFactory.java#L1788) is at this time. 

The reload will be invoked on a schedule setup within the provider here: https://github.com/airlift/airlift/pull/853/files#diff-06f4702cd9c78a864276eb020c71882aR32

Reply:
Thanks for explaining, I missed that part. In that case, there's no need for this to be a supplier -- it can be a utility method that returns the `SslContextFactory`. Having it be a supplier seems misleading and more complex.

With that change, this method can be eliminated and done similar to the one for the client:
```java
this.sslContextFactory = maybeSslContextFactory
        .orElseGet(() -> createReloadingSslContextFactory(config));
```
`SingletonSslContextFactorySupplier` will also no longer be needed.

Reply:
By misleading, I mean that `Supplier` is typically used so that consumers can repeatedly call `get()` to fetch a potentially different instance. Here, we only need to create it once, so there's no need for a supplier instance if we create it directly via a factory method.

Reply:
I'm not attached to using a Supplier, but I do think there are readability advantages to keeping the implementation of the class that manages the reloading of an SslContextFactory isolated from the implementation of HttpServer. 

The issue comes in how to implement the `reload` callback that the scheduled executor service invokes periodically. 

Here are a few examples I considered:

### Private instance method
There are a few subtle problems here. 

First, HttpServer has three allowed SslContextFactory states. It could be none/empty because Https isn't used, it could be the custom injected instance, or it could be the instantiated on construction SslContextFactory. Only the third one should be reloaded. This leaves open the possibility that reload will be invoked for either of the first two scenarios where it shouldn't be.  I don't _think_ this is that bad of a bug since reloading should generally be safe if it did happen. It is weird to read because if you aren't familiar with the complete context in which it's used it makes you wonder: "Why are we reloading when the SslContextFactory is not present?" 
```java
    private synchronized void reload()
    {
        try {
            SslContextFactory.Server updatedFactory = buildContextFactory(this.config);
            updatedFactory.start();
            if (this.sslContextFactory.isPresent()) {
                this.sslContextFactory.get().reload(factory -> factory.setSslContext(updatedFactory.getSslContext()));
            }
        }
        catch (Exception e) {
            log.warn(e, "Unable to reload SslContext.");
        }
    }
```

### Anonymous function
This one seems slightly better because it avoids the ambiguity around when reload should be invoked and what the allowable states of sslContextFactory are. It does bring up the question of how to best keep this thread safe and my inexperience with writing Java leaves me unsure of how best to handle it. I believe it requires some object to synchronize on, but I'm not sure which would be the best to use in this scenario. 
```java
 this.scheduledExecutorService.scheduleWithFixedDelay(() -> {
            synchronized (on something?) {
                try {
                    SslContextFactory.Server updatedFactory = buildContextFactory(config);
                    updatedFactory.start();
                    sslContextFactory.reload(factory -> factory.setSslContext(updatedFactory.getSslContext()));
                }
                catch (Exception e) {
                    log.warn(e, "Unable to reload SslContext.");
                }
            }
        }, refreshTime, refreshTime, MILLISECONDS);
```

My personal preference is to leave the the concern of reloading an SslContextFactory in a separate class and avoid this altogether. Changing the type to not implement Supplier may help a bit and then we are just left with a naming problem. If you feel strongly that its best to keep this inside the implementation of HttpServer, please let me know your thoughts on the above two suggestions. 

In the mean time, I'll work on addressing your other comments.

Reply:
I agree that having this in a separate class makes sense. My only concern is with implementing `Supplier`, so it's mostly just a naming (and usage) problem.

Reply:
I took a stab at a change, I now have it called a ReloadableSslContextFactoryProvider and added a few comments to explain it. It no longer implements Supplier. Still feels a bit off to me in naming but its the best I've come up with after sleeping on it. I'm open to suggestions if you can think of anything better.

Reply:
@electrum Pinging to see if you had any other feedback on this.

Reply:
@electrum Pinging again for any further feedback

Reply:
Sorry for the delay. This looks good to me.